go/tools/gopackagesdriver/archive_to_json.go

```diff
@@ -0,0 +1,187 @@
+	p.paramsReaderGetter = getter
+	return p
+}
+
+func (p *bazelArgsParser) WithArg(flag string, setter func(value string)) *bazelArgsParser {
+	arg := arg{
+		Setter:   setter,
+	}
+	p.args[flag] = arg
+	return p
+}
+
+func (p *bazelArgsParser) parseBazelArguments(scanner scanner) error {
+	var currentSetter func(value string)
+	for scanner.Scan() {
+		token := scanner.Text()
+		arg, ok := p.args[token]
+		if ok {
+			currentSetter = arg.Setter
+			continue
+		}
+		// We always expect to have a flag as first token.
+		if currentSetter == nil {
+			return errors.New(fmt.Sprintf("unexpected flag %s", token))
+		}
+		currentSetter(token)
+	}
+	return nil
+}
+
+func (p *bazelArgsParser) Parse(arguments []string) error {
+	// Nothing to do except validating.
+	if len(arguments) == 0 {
+		return nil
+	}
+	firstArgument := arguments[0]
+
+	// If the first argument is not a param argument,
+	// read the tokens from the arguments.
+	if firstArgument[0] != '@' {
+		scanner := newSliceScanner(arguments)
+		return p.parseBazelArguments(scanner)
+	}
+	if len(arguments) > 1 {
+		return errors.New("expected single argument with param file")
+	}
+
+	// We have a param argument, read tokens from the paramsReader
+	// using the param argument value.
+	paramsReader, err := p.paramsReaderGetter(firstArgument[1:])
+	if err != nil {
+		return err
+	}
+	scanner := bufio.NewScanner(paramsReader)
+	return p.parseBazelArguments(scanner)
+}
+
+func NewBazelArgsParser() *bazelArgsParser {
+	return &bazelArgsParser{
+		args: map[string]arg{},
+		paramsReaderGetter: func(value string) (io.Reader, error) {
+			return os.Open(value)
+		},
+	}
+}
+
+type archive struct {
+	ID              string   `json:"ID"`
+	PkgPath         string   `json:"PkgPath"`
+	ExportFile      string   `json:"ExportFile"`
+	GoFiles         []string `json:"GoFiles"`
+	CompiledGoFiles []string `json:"CompiledGoFiles"`
+	OtherFiles      []string `json:"OtherFiles"`
+}
+
+// Parse bazel generated arguments.
+// The generated arguments are expected to be in the multine format
+// for both the args and param file.
+func parseArchiveAndOutputPath(arguments []string, archive *archive) (string, error) {
+	var outputPath string
+	parser := NewBazelArgsParser().
+		WithArg("--id", func(value string) {
+			archive.ID = value }).
+		WithArg("--pkg-path", func(value string) {
+			archive.PkgPath = value }).
+		WithArg("--export-file", func(value string) {
+			archive.ExportFile = value }).
+		WithArg("--orig-srcs", func(value string) {
+			if strings.HasSuffix(value, ".go") {
+				archive.GoFiles = append(archive.GoFiles, value)
+			} else {
+				archive.GoFiles = append(archive.OtherFiles, value)
+			}
+		}).
+		WithArg("--data-srcs", func(value string) {
+			if strings.HasSuffix(value, ".go") {
+ 				archive.CompiledGoFiles = append(archive.CompiledGoFiles, value)
+ 			}
+		}).
+		WithArg("--output-file", func(value string) {
+			outputPath = value
+		})
```

Comment:
The argument parser implementation looks pretty good and there is no need to change it if it works, but I'm wondering whether you considered using https://pkg.go.dev/flag#FlagSet.Parse at some point.

Reply:
Yes, I used it initially! The problem I had with it is that I couldn't find a neat way to handle the multi valued args. I think bazel allows you to customize the output formats quite a lot, yet it felt easier to match what I wrote in the bazel aspect without having to format stuff to match go flags requirements.

Reply:
Anyways, I'm now testing it to check for potential issues and if it fixes my problem.